### PR TITLE
Replace text inputs with number stepper for project quota dialog

### LINKED ISSUE #4686: Update UI in quota dialog
According to new requirements, the text field in the project quota dialog needs to be updated to use a number stepper

- Current Implementation
image

- New Implementation
image

### PR DESCRIPTION
**Pull request by Talha-Jamil-TJ**
**What does this PR do / Why do we need it**:

This PR replaces `text inputs` with `number stepper inputs` for **project quota dialog**

> Before

image

> After
image

**Which issue(s) this PR fixes** :

Fixes #4686

**Release Note**:

```release-note
NONE
```

### CODE REVIEW

File: src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts [hunk 1]
```diff
@@ -58,6 +58,7 @@ export class ProjectQuotaDialogComponent implements OnInit, OnDestroy {
     private readonly _notificationService: NotificationService,
     private readonly _projectService: ProjectService,
     private readonly _quotaService: QuotaService,
+    private readonly _ref: ChangeDetectorRef,
```
Comment (Waseem826) [inline src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts:61]:
```suggestion
    private readonly _cdr: ChangeDetectorRef,
```

**Review by Waseem826 — CHANGES_REQUESTED** [review]
Some feedback. PTAL

**Comment by Waseem826** [comment]
We should replace other `RAM` words with `Memory`

screenshot-localhost_8000-2022 07 21-18_31_47

**Comment by Waseem826** [comment]
Can we just display project as a text instead of using a disabled mat-select?

screenshot-localhost_8000-2022 07 21-18_35_51

**Comment by Waseem826** [comment]
This is not related to this PR but would be nice if we can display `-` in empty columns.

screenshot-localhost_8000-2022 07 21-20_04_41

**Comment by Talha-Jamil-TJ** [comment]
> This is not related to this PR but would be nice if we can display `-` in empty columns.
> 
> screenshot-localhost_8000-2022 07 21-20_04_41

Need to get it approved from @cschieder

**Comment by Talha-Jamil-TJ** [comment]
> Can we just display project as a text instead of using a disabled mat-select?
> 
> screenshot-localhost_8000-2022 07 21-18_35_51

As there wasn't any designs for this modal, I reused the `add` dialog for `edit`, and disabling the project select was suggested by @lsviben

File: src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts (hunk 1 above)
Comment (Talha-Jamil-TJ) [reply src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts:61]: feedback implemented

File: src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts [hunk 2]
```diff
@@ -168,5 +169,19 @@ export class ProjectQuotaDialogComponent implements OnInit, OnDestroy {
         this.selectedProject = this.projects.find(({id}) => id === projectId);
         this.selectedQuota = this._quotas.find(({subjectName}) => subjectName === projectId);
       });
+
+    // This subscription prevents "Expression changed after checked" error
```
[the remark this answers is not in the document]
Comment (Talha-Jamil-TJ) [reply src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts:173]: feedback implemented

[the remark this answers is not in the document]
Comment (Talha-Jamil-TJ) [reply src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts:173]: @Waseem826 what about the comment  on the above subscription?

File: src/app/dynamic/enterprise/quotas/quota.ts [hunk 3]
```diff
@@ -18,7 +18,7 @@ import {Observable, merge, timer, Subject} from 'rxjs';
 import {switchMap, shareReplay, map, takeUntil} from 'rxjs/operators';
 import {environment} from '@environments/environment';
 import {QuotaDetails, Quota, QuotaVariables} from '@shared/entity/quota';
-import {AppConfigService} from '../../config.service';
+import {AppConfigService} from '../../../config.service';
 
 @Injectable()
 export class QuotaService implements OnDestroy {
```
[the remark this answers is not in the document]
Comment (Talha-Jamil-TJ) [reply src/app/dynamic/enterprise/quotas/quota.ts:24]: file renamed to `service.ts`

File: src/app/shared/components/number-stepper/template.html [hunk 4]
```diff
@@ -28,6 +28,7 @@
          [min]="min"
          [max]="max"
          [id]="id"
+         (blur)="onTouch()"
```
[the remark this answers is not in the document]
Comment (Talha-Jamil-TJ) [reply src/app/shared/components/number-stepper/template.html:31]: According to angular docs this should be `blur` event.

image

see: https://angular.io/api/forms/ControlValueAccessor

File: src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Waseem826) [reply src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts:173]: We can keep that

File: src/app/dynamic/enterprise/quotas/component.ts [hunk 5]
```diff
@@ -116,25 +116,28 @@ export class QuotasComponent implements OnInit {
...
+        confirmLabel: 'Delete',
+      },
+    } as MatDialogConfig;
+
+    this._matDialog
+      .open<ConfirmationDialogComponent, ConfirmationDialogConfig, boolean>(ConfirmationDialogComponent, config)
+      .afterClosed()
+      .pipe(
+        filter(confirmed => confirmed),
+        switchMap(_ => this._quotaService.deleteQuota(name))
```
[the remark this answers is not in the document]
Comment (Talha-Jamil-TJ) [reply src/app/dynamic/enterprise/quotas/component.ts:135]: we don't need to manually unsubscribe to `afterClosed()` as it does it on its own

image

File: src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts [hunk 6]
```diff
@@ -168,5 +169,19 @@ export class ProjectQuotaDialogComponent implements OnInit, OnDestroy {
...
+    this.quotaGroup.statusChanges.pipe(takeUntil(this._unsubscribe)).subscribe(_ => this._ref.detectChanges());
+
+    // This subscription will set quota inputs to empty state when user goes below 1
+    Object.values(this.quotaGroup.controls).forEach(control => {
```
[the remark this answers is not in the document]
Comment (Talha-Jamil-TJ) [reply src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts:177]: Yes, users can enter `0`, but it won't make the form valid, and those values won't go into the API payload.

**Review by Waseem826 — APPROVED** [review]
/approve

**Pull request merged**